pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -413,27 +415,127 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
     }
 
     String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(segmentMetadata.getTableName());
-    OfflineSegmentZKMetadata offlineSegmentZKMetadata = _pinotHelixResourceManager.getOfflineSegmentZkMetadata(
-        offlineTableName, segmentName);
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(offlineTableName, segmentName);
 
-    if (offlineSegmentZKMetadata == null) {
+    if (znRecord == null) {
       // Creating new segment, not refreshing
       downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
       _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
       response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
       return response;
     }
-    if (downloadUrl == null) {
-      downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
+
+    final OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
+
+    long minionExpectedCrc = -1L;
+    boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
+    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    boolean metadataRewritten = false;
+    if (isMinionConfigured) {
+      // Minion configured use case so that we write to zk sparingly
+      String segmentPushStatus = originalOfflineSegmentZkMetadata.getSegmentPushStatus();
+      SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
+      Date dt = new Date();
```

Comment:
Make sure this is the same as system time.

Reply:
Good call. From the docs: /**
     * Allocates a <code>Date</code> object and initializes it so that
     * it represents the time at which it was allocated, measured to the
     * nearest millisecond.
     *
     * @see     java.lang.System#currentTimeMillis()
     */